app/src/main/java/org/mozilla/rocket/download/DownloadInfoViewModel.kt

```diff
@@ -0,0 +1,264 @@
+                repository.remove(download.rowId)
+            } else {
+                toastMessage.value = R.string.cannot_delete_the_file
+            }
+        } catch (e: Exception) {
+            Log.e(this.javaClass.simpleName, "" + e.message)
+            toastMessage.value = R.string.cannot_delete_the_file
+        }
+    }
+
+    fun add(downloadInfo: DownloadInfo) {
+        var index = -1
+        for (i in 0 until downloadInfoBundle.list.size) {
+            if (downloadInfoBundle.list[i].rowId < downloadInfo.rowId) {
+                index = i
+                break
+            }
+        }
+
+        if (index == -1) {
+            downloadInfoBundle.list.add(downloadInfo)
+            //The crash will happen when data set size is 1 after add item.
+            //Because we define item count is 1 and mDownloadInfo is empty that means nothing and show empty view.
+            //So use notifyDataSetChanged() instead of notifyItemInserted when data size is 1 after add item.
+            if (downloadInfoBundle.list.size > 1) {
+                downloadInfoBundle.notifyType = DownloadInfoBundle.Constants.NOTIFY_ITEM_INSERTED
+                downloadInfoBundle.index = (downloadInfoBundle.list.size - 1).toLong()
+            } else {
+                downloadInfoBundle.notifyType = DownloadInfoBundle.Constants.NOTIFY_DATASET_CHANGED
+            }
+        } else {
+            downloadInfoBundle.list.add(index, downloadInfo)
+            downloadInfoBundle.notifyType = DownloadInfoBundle.Constants.NOTIFY_ITEM_INSERTED
+            downloadInfoBundle.index = index.toLong()
+        }
+        downloadInfoList.value = downloadInfoBundle
+    }
+
+    fun hide(rowId: Long) {
+        for (i in 0 until downloadInfoBundle.list.size) {
+            val downloadInfo = downloadInfoBundle.list[i]
+            if (rowId == downloadInfo.rowId) {
+                downloadInfoBundle.list.remove(downloadInfo)
+                downloadInfoBundle.notifyType = DownloadInfoBundle.Constants.NOTIFY_ITEM_REMOVED
+                downloadInfoBundle.index = i.toLong()
+                downloadInfoList.value = downloadInfoBundle
+                break
+            }
+        }
+    }
+
+    fun updateRunningItems() {
+        val runningIds = runningDownloadIds
+        if (!runningIds.isEmpty()) {
+            for (i in runningDownloadIds.indices) {
+                repository.queryByDownloadId(runningDownloadIds[i], updateListener)
+            }
+        }
+    }
+
+    fun notifyDownloadComplete(downloadId: Long) {
+        repository.queryByDownloadId(downloadId, updateListener)
+    }
+
+    fun notifyRowUpdate(rowId: Long) {
+        repository.queryByRowId(rowId, updateListener)
+    }
+
+    private inner class QueryProgressRunnable : Runnable {
+        override fun run() {
+            if (queryDownloadProgress()) {
+                handler?.postDelayed(this, QUERY_PROGRESS_DELAY)
```

Comment:
This implementation might stack up more messages if Query takes more than QUERY_PROGRESS_DELAY. If this happens, there will be a race condition between queries of queryDownloadItems. I think this is quite unlikely given QUERY_PROGRESS_DELAY = 500 but we might want to check if we have some time. Also since these will be cleaned up so the impact is not that severe.